File: pkg/funcreg/src/linfda.f
c     ***********************************************************
c     ***********************************************************
c                             LINEAR FDA
c     ***********************************************************
c     ***********************************************************      
      
c     Codes to estimate the coefficients for linear fda
c     ***********************************************************
      subroutine coefest(y, basisval, n, k, t, lambda, pen, coef)
      integer n, k, t, p(k), infotr, infosv
      double precision basisval(t,k), lambda, coef(k,n)
      double precision y(t,n), pen(k,k), tbasis(k,t), pen2(k,k)
      
      pen2 = pen
      call dgemm('T', 'N', k, k, t, 1.0d0, basisval, t, basisval, t,
     *     lambda*t, pen2, k)
      call dgetrf(k, k, pen2, k, p, infotr)
      tbasis = transpose(basisval)
      call dgetrs('N', k, t, pen2, k, p, tbasis, k, infosv)
      
      call dgemm('N', 'N', k, n, t, 1.0d0, tbasis, k, y, t,
     *     0.0d0, coef, k)

      end

c     Codes to get the coefficients and yhat's for linear fda
c     and observation i removed
c     ***********************************************************
      subroutine yhati(y, basisval, n, k, t, lambda, pen, i, coef, yhat)
      integer n, k, t, s(t-1), i, j
      double precision basisval(t,k), lambda, coef(k,n)
      double precision y(t,n), pen(k,k), yhat(n)

      do j=1,(t-1)
         if (j < i) then
            s(j) = j
         else
            s(j) = j+1
         end if
      end do      
      call coefest(y(s,:), basisval(s,:), n, k, t-1, lambda, pen,
     *     coef)
      call dgemv('T', k, n, 1.0d0, coef, k, basisval(i,:), 1, 0.0d0,
     *     yhat, 1)
   
      end 


c     Codes to get the leave-one-out cross-validation for linear fda
c     lambda can be a vector, which will produce a vector of CV.
c     ***********************************************************      
      subroutine crval(y, basisval, n, k, t, nlam, lambda, pen, cv)
      integer n, k, t, i, j, nlam
      double precision y(t,n), basisval(t,k), lambda(nlam), pen(k,k)
      double precision cv(nlam), coef(k,n), yhat(n)
      cv = 0.0d0
      do j=1,nlam
         do i=1,t
            call yhati(y, basisval, n, k, t, lambda(j), pen, i, coef,
     *           yhat)
            cv(j) = cv(j) + sum((yhat-y(i,:))**2)
         end do
      end do
      
      end

c     Minimization of the cross-validation with respect to
c     lambda for a given k using the Golden-Section method
c     The linear case. 
c     ********************************************************* 
      subroutine lamgs(y, basisval, n, k, t, l1, l2, 
     *     pen, maxitalgo, tolalgo, info, iter, lam, cvf)
      integer n, t, k, maxitalgo, i, info, iter
      double precision y(t,n), basisval(t,k), l1, l2
      double precision lam, pen(k,k), tmpl(2), cvf
      double precision cv(2), gspar, tolalgo

      gspar = 1-0.3819660
      tmpl(1) = gspar*l1+(1-gspar)*l2
      tmpl(2) = (1-gspar)*l1+gspar*l2
      info = 0
      i = 1
      call crval(y, basisval, n, k, t, 2, tmpl, pen, cv)
      do
         iter = i
         if (cv(1) < cv(2)) then
            l2  = tmpl(2)
            tmpl(2) = tmpl(1)
            cv(2) = cv(1)
            lam = tmpl(2)
            cvf = cv(2)
            tmpl(1) = gspar*l1+(1-gspar)*l2
            call crval(y, basisval, n, k, t, 1, tmpl(1), pen, cv(1))
         else
            l1 = tmpl(1)
            tmpl(1) = tmpl(2)
            lam = tmpl(1)
            cv(1) = cv(2)
            cvf = cv(1)
            tmpl(2) = (1-gspar)*l1+gspar*l2
            call crval(y, basisval, n, k, t, 1, tmpl(2), pen, cv(2))
         end if
         if (abs(l2-l1) < lam*tolalgo) then
            exit
         end if
         if (i>maxitalgo) then
            info = 1
            exit
         end if
         i = i+1
      end do
      end

c     Minimization of the cross-validation with respect to
c     lambda for a given k using the Brent method
c     The linear case. 
c     *********************************************************       
      subroutine lambrent(y, basisval, n, k, t, ax, cx, 
     *     pen, maxita, tola, info, iter, x, fx)
      integer n, t, k, maxita, i, info, iter
      double precision y(t,n), basisval(t,k), l1, l2
      double precision pen(k,k), tmpl(2), ax, cx, bx
      double precision cv(2), tola, tol1, tol2
      double precision gs, zeps, a, b, d, etemp, fu, fv, fw, fx
      double precision p,q,r,u,v,w,x,xm,e,u2(1), fu2(1)
      gs = 0.3819660
      e = 0.0d0
      zeps = epsilon(gs)*1.0e-3
c     Find the third point between ax and cx
      tmpl(1) = (1-gs)*ax+gs*cx
      tmpl(2) = gs*ax+(1-gs)*cx
      call crval(y, basisval, n, k, t, 2, tmpl, pen, cv)
      if (cv(1) < cv(2)) then
         bx = tmpl(1); cx = tmpl(2); fx = cv(1)
      else
         bx = tmpl(2); ax = tmpl(1); fx = cv(2)
      end if
c     ############################
      x = bx; w = bx; v = x; fv = fx; fw = fx
      i = 1; a = ax; b = cx; info = 0
      do
         iter = i
         if (i>maxita) then
            info = 1
            exit
         end if
         xm = 0.5*(b+a)
         tol1 = tola*abs(x)+zeps
         tol2 = 2.0*tol1
         if (abs(x-xm) <= (tol2-0.5*(b-a))) then
            exit
         end if
         if (abs(e) > tol1) then
            r = (x-w)*(fx-fv)
            q = (x-v)*(fx-fw)
            p = (x-v)*q-(x-w)*r
            q = 2.0*(q-r)
            if (q>0.0) then
               p = -p
            end if
            q = abs(q)
            etemp = e
            e = d
            if (abs(p) >= abs(0.5*q*etemp) .or. p <= q*(a-x) .or.
     *           p >= q*(b-x)) then
               if (x >= xm) then
                  e = a-x
               else
                  e = b-x
               end if               
               d = gs*e
            else
               d = p/q
               u = x+d
               if (u-a < tol2 .or. b-u < tol2) then
                  d = sign(tol1,xm-x)
               end if
            end if   
         else
            if (x >= xm) then
               e = a-x
            else
               e = b-x
            end if               
            d = gs*e
         end if
         if (abs(d) >= tol1) then
            u=x+d
         else
            u = x+sign(tol1,d)
         end if
         u2(1) = u
         call crval(y, basisval, n, k, t, 1, u2, pen, fu2)
         fu = fu2(1)
         if (fu <= fx) then
            if (u >= x) then
               a=x
            else
               b=x
            end if
            v = w; w = x; x = u; fv = fw; fw = fx; fx = fu
         else
            if (u<x) then
               a=u
            else
               b=u
            end if
            if (fu <= fw .or. w == x) then
               v = w; w = u; fv = fw; fw = fu
            else if (fu <= fv .or. v == x) then
               v = u; fv = fu
            end if
         end if
         i = i+1
      end do
      end

      

File: pkg/funcreg/R/linFda.R
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  A copy of the GNU General Public License is available at
#  http://www.r-project.org/Licenses/



#### Function to get the coefficients
#### Y is a matrix, not a list (txn)

coefEst <- function(y, t, lam, k, L=2,
                    create_basis=create.bspline.basis, ...)
    {
        type <- "Linear FDA Fit"
        y <- as.matrix(y)
        n <- ncol(y)
        T <- nrow(y)    
        if (length(t) != T)
            stop("nrow(y) must be equal to length(t)")
        rangeval <- range(t)
        basis <- create_basis(rangeval,k, ...)
        basisval <- eval.basis(t,basis)
        pen <- eval.penalty(basis,L)    
        res <- .Fortran("coefest", as.double(y), as.double(basisval), as.integer(n),
                        as.integer(k), as.integer(T), as.double(lam), as.double(pen),
                        coef = double(k*n), PACKAGE="funcreg")
        ans <- list(coefficients=matrix(res$coef, k, n), convergence=NULL,
                    t=t, y=y, basis=basis, link=function(x) x,
                    lambda=lam, L=L, type=type)
        class(ans) <- "myfda"
        ans
    }

#### Main function to get the CV matrix (nlam x nK)
#### lam and k are vectors
####################################################
fdaCV <- function(y, t, lamvec, kvec, L=2,
                  create_basis=create.bspline.basis, obj=NULL, ...)
    {
        if (!is.null(obj))
            {
                if (class(obj) != "myfda")
                    stop("object must be of class myfda")
                y <- obj$y; t <- obj$t; kvec <- obj$basis$nbasis; n <- ncol(y)
                L <- obj$L; T <- nrow(y); nlam <- 1; lamvec <- obj$lambda
            } else {
                y <- as.matrix(y)
                n <- ncol(y)
                T <- nrow(y)
                nlam <- length(lamvec)
                if (length(t) != T)
                    stop("nrow(y) must be equal to length(t)")
            }
        rangeval <- range(t)
        f <- function(k, basis=NULL)
            {
                if (is.null(basis))
                    basis <- create_basis(rangeval,k, ...)
                basisval <- eval.basis(t,basis)
                pen <- eval.penalty(basis,L)    
                res <- .Fortran("crval", as.double(y), as.double(basisval),
                                as.integer(n), as.integer(k), as.integer(T),
                                as.integer(nlam), as.double(lamvec),
                                as.double(pen), cv = double(nlam), PACKAGE="funcreg")
                res$cv
            }
        if (is.null(obj))
            ans <- sapply(kvec, f)
        else
            ans <- f(kvec, obj$basis)
        ans
    }

## This function finds the optimal Lambda for a given k
## using either the Brent method or the Golden Section method
## maxitalgo and tolalgo are the parameters for the minimization of the
## cross-validation. The linear Case
 #################################################
getLam <- function(y, t, lamInt, k, L=2,
                   create_basis=create.bspline.basis, 
                   maxitalgo=100, tolalgo=1e-7, type=c("Brent","GS"), ...)
    {
        type = match.arg(type)
        type = ifelse(type=="Brent", "lambrent", "lamgs")
        rangeval <- range(t)
        basis <- create_basis(rangeval,k, ...)
        basisval <- eval.basis(t,basis)
        pen <- eval.penalty(basis,L)    
        T <- nrow(y)
        n <- ncol(y)
        res <- .Fortran(type, as.double(y), as.double(basisval),
                        as.integer(n), as.integer(k), as.integer(T),
                        as.double(lamInt[1]), as.double(lamInt[2]), as.double(pen),
                        as.integer(maxitalgo), as.double(tolalgo), 
                        info = integer(1), iter = integer(1),
                        lam = double(1), cv=double(1), PACKAGE="funcreg")
        c(lam=res$lam, info=res$info, iter=res$iter, cv=res$cv)                        
    }


### This function search over kvec and
### pick the one with minimum CV. For each K, the optimal Lambda is obtained
### numerically either by the Brent Method or the Golden Section Method.
### maxitalgo and tolalgo are the parameters for the minimization of the
### cross-validation. logLamInt is a 2x1 vector with the upper and lower bound
### for the search.
### It returns an object of class myfda
### The linear case
#########################################################################
getLandKopt <- function(y, t, lamInt, kvec, L=2,
                        create_basis=create.bspline.basis, 
                        maxitalgo=100, tolalgo=1e-6, type=c("Brent", "GS"), ...)
    {
        type <- match.arg(type)
        res <- sapply(kvec, function(k) getLam(y=y, t=t, lamInt,
                                               k=k, L=L,
                                               create_basis=create_basis,
                                               maxitalgo=maxitalgo,
                                               tolalgo=tolalgo, type=type, ...))
        w <- which(res[4,]==min(res[4,],na.rm=TRUE))
        lambda <- res[1,w]
        k=kvec[w]
        obj <- coefEst(y=y, t=t, lam=lambda, k=k, L=L,
                       create_basis=create_basis, ...)
        obj$cv <- res[4,w]
        obj$algo <- list(lamInter=lamInt, convergence=res[2,], numIter=res[3,])
        obj
    }

### Function to get the minimum CV
#####################################
getLandK <- function(y, t, lamvec, kvec, L=2,
                     create_basis=create.bspline.basis, ...)
    {
        cv <- fdaCV(y, t, lamvec, kvec, L, create_basis, ...)
        w <- which(cv==min(cv,na.rm=TRUE), arr.ind=TRUE)
        lambda <- lamvec[w[1]]
        k=kvec[w[2]]
        res <- coefEst(y=y, t=t, lam=lambda, k=k, L=L,
                         create_basis=create_basis, ...)
        res$cv <- min(cv,na.rm=TRUE)
        res$grid <- list(lamgrid=lamvec, kgrid=kvec, cv=cv)
        res
    }

